concent_api/middleman/queue_operations.py

```diff
@@ -0,0 +1,196 @@
+from asyncio import IncompleteReadError
+from asyncio import Queue
+from asyncio import StreamReader
+from asyncio import StreamWriter
+from collections import OrderedDict
+from logging import getLogger
+from logging import Logger
+from typing import Dict
+
+from django.conf import settings
+from django.utils.timezone import now
+
+from middleman.constants import CONNECTION_COUNTER_LIMIT
+from middleman.constants import MessageTrackerItem
+from middleman.constants import RequestQueueItem
+from middleman.constants import ResponseQueueItem
+from middleman_protocol.constants import PayloadType
+from middleman_protocol.constants import REQUEST_ID_FOR_RESPONSE_FOR_INVALID_FRAME
+from middleman_protocol.exceptions import BrokenEscapingInFrameMiddlemanProtocolError
+from middleman_protocol.exceptions import FrameInvalidMiddlemanProtocolError
+from middleman_protocol.exceptions import MiddlemanProtocolError
+from middleman_protocol.exceptions import PayloadTypeInvalidMiddlemanProtocolError
+from middleman_protocol.exceptions import SignatureInvalidMiddlemanProtocolError
+from middleman_protocol.message import ErrorFrame
+from middleman_protocol.message import GolemMessageFrame
+from middleman_protocol.registry import create_middleman_protocol_message
+from middleman_protocol.stream_async import handle_frame_receive_async
+from middleman_protocol.stream_async import map_exception_to_error_code
+from middleman_protocol.stream_async import send_over_stream_async
+
+logger = getLogger()
+
+
+async def request_producer(
+    request_queue: Queue,
+    response_queue: Queue,
+    reader: StreamReader,
+    connection_id: int,
+) -> None:
+    while True:
+        try:
+            frame = await handle_frame_receive_async(reader, settings.CONCENT_PUBLIC_KEY)
+        except IncompleteReadError:
+            logger.info(f"Client has closed the connection: {connection_id}. Ending 'request_producer' coroutine")
+            break
+        except (FrameInvalidMiddlemanProtocolError,
+                BrokenEscapingInFrameMiddlemanProtocolError,
+                SignatureInvalidMiddlemanProtocolError,
+                PayloadTypeInvalidMiddlemanProtocolError) as exception:
+            logger.info("Received invalid message")
+            await response_queue.put(
+                ResponseQueueItem(
+                    create_error_frame(exception),
+                    REQUEST_ID_FOR_RESPONSE_FOR_INVALID_FRAME,
+                    now().timestamp()
```

Comment:
Why not reusing `get_current_utc_timestamp` ?

Reply:
Yeah. The timestamps are very tricky due to time zones. There were bugs related to it both in our code and in Golem. Ultimately we have created a set of helpers for creating and converting them and a rule never to do it directly. 

Even if you think you have it right you should always use them so that the person reading the code does not even have to think about it.

Reply:
done